Manually kill neutron-ns-metadata-proxy

nkrinner opened the pull request:
...persist

when we shut down neutron, as is needed during the upgrade process from Cloud 4
to Cloud 5. This will break the upgrade. Upstream iss working on the handling
of external processes, but in the meantime we need a workaround.

Get nodes with the neutron-l3 role and manually kill the neutron-ns-metadata-proxy
process during the upgrade.

nkrinner commented:
We could also fix the problem in the respective openstack-neutron init script. Technically we could run into the problem of a unstopped neutron-ns-metadata-proxy process again else wehere. It is a matter of taste I think.

aspiers commented:
I don't know much about this proxy process but instinctively I would much prefer to fix it in the init script - like you say this problem is not exclusive to upgrade, so it seems wrong to me to fix it only for upgrade.

But I could be missing something.

BTW please keep the first line of commit messages short (ideally <= 50 characters but definitely < 80).

rossella commented:
The metadata proxy process is spawned by the l3 agent to make the metadata server reachable by the VM and to attach the proper data (instance_id, network_id, etc) in the http request. It's probably not disruptive to move the killall in the init script, since all those proxies are spawned by the l3 agent, anyway I'd go with the safer approach of fixing it in the upgrade script. Till now only the upgrade process is showing the bug. If we add the killall in the init script we might break something else.

rossella commented on lib/suse-cloud-upgrade-4-to-5-pre:
```diff
@@ -205,6 +205,21 @@ for node in $allocated_nodes; do
+### Workaround for neutron shortcoming
+# Currently neutron-ns-metadata-proxy is not stoppend on a HA setup. The running process
+# kills the update, so we need to kill it manually. Upstream is working on improving the
+# handling of external processes. Once this is achieved we don't need this part anymore
```
Indeed there's a blueprint approved for that [1] .
Anyway the blueprint doesn't take care of killing orphaned external processes. This can be proposed upstream. So I guess that for now we have to do it manually.

[1] https://blueprints.launchpad.net/neutron/+spec/agent-child-processes-status

jsuchome commented:
+1

vuntz commented on lib/suse-cloud-upgrade-4-to-5-pre:
```diff
@@ -205,6 +205,21 @@ for node in $allocated_nodes; do
+# Currently neutron-ns-metadata-proxy is not stoppend on a HA setup. The running process
+# kills the update, so we need to kill it manually. Upstream is working on improving the
+# handling of external processes. Once this is achieved we don't need this part anymore
+get_allocated_nodes neutron-l3
```
This won't work: we've disabled the neutron proposal, so this will return nothing ;-)

So either get the list of l3 nodes earlier, or just run this killall bit on all nodes.

jsuchome replied:
So maybe just call killall during the previous ```for``` loop, it doesn't hurt that not all the nodes have neutron installed

vuntz commented on lib/suse-cloud-upgrade-4-to-5-pre:
```diff
@@ -205,6 +205,21 @@ for node in $allocated_nodes; do
+  echo "Killing neutron-ns-metadata-proxy process on $node..."
+  ssh "$node" '
+    killall neutron-ns-metadata-proxy 
+    '
```
You can use one line instead of three lines here :-)

nkrinner commented:
Updated the pull request according to the pull request. There was the suggestion to add the killall part to the loop where we disable the openstack services. I kept it separate, as I think it is just a workaround and it will stay more visible this way. And we probably want to get rid of it again at some later point.

I don't object changing this, if people prefer it done in the other way.

Also changed the commit message so the first line is now less  than 50 chars and more on point.

vuntz commented:
It was already mentioned by @aspiers in https://github.com/SUSE-Cloud/suse-cloud-upgrade/pull/34#issuecomment-71846571 : your first commit line is way too long.

vuntz commented:
+1 on the code, but the commit message should really be fixed.

nkrinner commented:
@vuntz I think the commit message was already fixed with my last commit. I also adapted the comment in the pull request that was created from the original commit message. I hope this is better now?

vuntz commented:
Ah sorry, the pull request title was what I saw
